ui/src/treegrid.tsx

```diff
@@ -0,0 +1,247 @@
+import ErrorIcon from "@mui/icons-material/Error";
+import KeyboardArrowDownIcon from "@mui/icons-material/KeyboardArrowDown";
+import KeyboardArrowRightIcon from "@mui/icons-material/KeyboardArrowRight";
+import CircularProgress from "@mui/material/CircularProgress";
+import IconButton from "@mui/material/IconButton";
+import Typography from "@mui/material/Typography";
+import { ReactNode, useCallback } from "react";
+import { useImmer } from "use-immer";
+
+export type TreeGridId = string | number;
+
+export type TreeGridRowData = {
+  [key: string]: string | number | JSX.Element;
+};
+
+export type TreeGridItem = {
+  children?: TreeGridId[];
+  data: TreeGridRowData;
+};
+
+export type TreeGridData = {
+  [key: TreeGridId]: TreeGridItem;
```

Comment:
The caller cannot reliably impose an order on the rows in this model (except maybe with certain browsers and a deep understanding of object key iteration order). Is that OK? Something like `Map` would preserve insertion order.